.buildkite/pipeline.yml

```
@@ -57,18 +64,9 @@ steps:
     plugins: *common_plugins
 
   - label: "Instrumented tests"
-    key: "instrumented_tests"
     command: .buildkite/commands/run-instrumented-tests.sh
-    plugins: *common_plugins
+    plugins: *common_plugins_with_test_collector
+    env:
+      BUILDKITE_ANALYTICS_TOKEN: $BUILDKITE_ANALYTICS_TOKEN_INSTRUMENTED_TESTS  
```

Have you tried setting the `api-token-env-name` variable? This variable [should take precedence](https://github.com/buildkite-plugins/test-collector-buildkite-plugin/blob/634bc81454108ca01df7888bbe0eceadaaafee2c/hooks/pre-exit#L4C9-L4C9) over `BUILDKITE_ANALYTICS_TOKEN`.
```suggestion
    env:
      BUILDKITE_PLUGIN_TEST_COLLECTOR_API_TOKEN_ENV_NAME: BUILDKITE_ANALYTICS_TOKEN_INSTRUMENTED_TESTS
```

---

👋 @crazytonyli 

> Have you tried setting the api-token-env-name variable?

I just tried that (in another draft PR) and it works. Could you please explain what the difference between using these two, security-wise?

The thing I was concerned about in general was referencing secrets in the pipeline file, which is considered an antipattern [here](https://buildkite.com/docs/pipelines/secrets#anti-pattern-referencing-secrets-in-your-pipeline-yaml). Technically, this is an antipattern, however, I'm not sure if this is an actual security concern. I just wanted to double-check if it's fine, because I don't have such a context as you do.

---

I wouldn't suggest to use `BUILDKITE_PLUGIN_TEST_COLLECTOR_API_TOKEN_ENV_NAME: …` as `env` as @crazytonyli suggested above, because those `BUILDKITE_PLUGIN_<PLUGIN_NAME>_<CONFIGURATION_PROPERTY>` env vars are special env vars used within Buildkite Plugin implementations, to expose the the plugins' properties. So they are not meant to be set manually, but instead are set by Buildkite's plugin architecture when you use a plugin in your pipeline and pass it some parameters.

So in this case, this `BUILDKITE_PLUGIN_TEST_COLLECTOR_API_TOKEN_ENV_NAME` env var is set by Buildkite with the value [you pass to the `api-token-env-name` parameter of the plugin](https://github.com/buildkite-plugins/test-collector-buildkite-plugin/blob/634bc81454108ca01df7888bbe0eceadaaafee2c/plugin.yml#L8)

This actually means that, from what I understand from [that `test-collector` plugin's definition file](https://github.com/buildkite-plugins/test-collector-buildkite-plugin/blob/634bc81454108ca01df7888bbe0eceadaaafee2c/plugin.yml#L7-L8), the way you're expected to pass the api token is not via passing the value via `env`, but instead to pass the env var name as the `api-token-env-name` property of your plugin:

```yaml
    - test-collector#v1.8.0:
        files: "WooCommerce/build/buildkite-test-analytics/*.xml"
        format: "junit"
        api-token-env-var: BUILDKITE_ANALYTICS_TOKEN_INSTRUMENTED_TESTS
```
(Note how this doesn't use `$`, as this is supposed to be just the name of the env var itself, not its value. Which is also how this protects from the variable to be interpolated inside your pipeline and risking exposing your secret (which you were rightfully concerned about).

---

> The thing I was concerned about in general was referencing secrets in the pipeline file, which is considered an antipattern [here](https://buildkite.com/docs/pipelines/secrets#anti-pattern-referencing-secrets-in-your-pipeline-yaml). I'm not sure if this is an actual security concern, though, but wanted to double-check, because I don't have such a context as you do.

The anti-pattern that is described here in the doc is a good thing to have in mind, but it applies to the risk of the env vars being interpolated. So what it's saying is that the thing to avoid at all cost is to use `command: "mytool --token "$MY_TOKEN"` as a command on your steps in your `pipeline.yml`, because that `$MY_TOKEN` will be interpolated and replaced with the actual token / secret, and then will be visible in logs (and exposed as part of the `BUILDKITE_COMMAND` env var that Buildkite set with the command being run by your step).

This is why they recommend instead to wrap such calls into a shell script (e.g. have a `.buildkite/commands/run-mytool.sh` script whose content is `mytool --token $MY_TOKEN`, and have your steps' `command` be `command: "./buildkite/commands/run-mytool.sh` instead, so that the token is not exposed (and at the risk of being interpolated) directly in your pipeline, but hidden away inside a script (whose content isn't exposed in the logs or in the `BUILDKITE_COMMAND` env var)

I'm not entirely sure of that rule applies to `env: { TOKEN: $MY_TOKEN }` though, especially as Buildkite automatically redacts the content of env vars whose name ends with `_TOKEN` and similar magic words.
But the fact that the `test-collector` plugin explicitly exposes an `api-token-env-var` property makes me pretty confident that this is how they expect you to pass the token anyway — especially since that property only takes the _name_ of that token env var, not its value, so there's no risk of your pipeline interpolating and exposing that value direclty.

---

PS: Unrelated to your question, but while I was looking at the plugin's README to answer this, I noticed [this other optional parameter](https://github.com/buildkite-plugins/test-collector-buildkite-plugin/blob/main/README.md#annotation-linkboolean), which means that if you are interested, you could pass `annotation-link: true` to the plugins' properties too (alongside `files`, `format` and `api-token-env-var`) to let the plugin add a Buildkite annotation containing the link to the test report. Could be cool and useful 🙂

---

Thank you so much @AliSoftware 🙇 It will take me some time to try the things you've suggested, and get back here with the results. This was very helpful!

---

@AliSoftware Currently, in the trunk branch, the pipeline is [using the plugin property to pass token](https://github.com/woocommerce/woocommerce-android/blob/trunk/.buildkite/pipeline.yml#L71-L74), as you see in this PR diff. I believe what @pachlava wanted to do in this PR is adding the same plugin to another test step, but with a different `api-token-env-var` value. See https://github.com/woocommerce/woocommerce-android/pull/9380#discussion_r1254645343

---

> I believe what @pachlava wanted to do in this PR is adding the same plugin to another test step, but with a different api-token-env-var value.

This is correct @crazytonyli. For Woo we need at least two different tokens, and for WP/JP up to five (UI test for iPhone, iPad, and three kinds of unit tests).

Before the suggestions from @AliSoftware, this was the approximate code used for this (already using the plugin declaration improvement suggested by @AliSoftware):

```yml
  - label: "Instrumented tests"
    command: .buildkite/commands/run-instrumented-tests.sh
    plugins: [*ci_toolkit, *test_collector]
    env:
      BUILDKITE_ANALYTICS_TOKEN: $BUILDKITE_ANALYTICS_TOKEN_INSTRUMENTED_TESTS
    artifact_paths:
      - "**/build/instrumented-tests/**/*"
```

However, I was not sure of the fact of referencing `$BUILDKITE_ANALYTICS_TOKEN_INSTRUMENTED_TESTS`, and this was one of the suggestions from @AliSoftware:

> This is why they recommend instead to wrap such calls into a shell script (e.g. have a .buildkite/commands/run-mytool.sh script whose content is mytool --token $MY_TOKEN, and have your steps' command be command: "./buildkite/commands/run-mytool.sh instead, so that the token is not exposed

With a bit of different notation, this was the first thing I tried in one of the other earlier PRs, that added analytics for Instrumented Tests first. Similarly to how it was [done in WCiOS](https://github.com/woocommerce/woocommerce-ios/pull/7490/files#diff-bd954ba2ef49cba9cd9950b5197be8e22f9c95f5482f03d0186eaf9404cda0e8), I've used `export BUILDKITE_ANALYTICS_TOKEN=$BUILDKITE_ANALYTICS_TOKEN_INSTRUMENTED_TESTS` inside `run-instrumented-tests.sh`, but it always resulted in `Missing BUILDKITE_ANALYTICS_TOKEN environment variable`, because, as I see it, the var will be set in the shell script process, not in Buildkite command process. This stays true for the case of using the `environment` hook [as suggested here](https://buildkite.com/docs/pipelines/secrets#exporting-secrets-with-environment-hooks). In fact, the current state of this PR experiments with this (for Unit Tests only).

> mytool --token $MY_TOKEN

I could not get this notation working (tbh, I don't understand what should I use for `mytool`), could you please explain it in more details, @AliSoftware? Thank you!

---

> @AliSoftware Currently, in the trunk branch, the pipeline is [using the plugin property to pass token](https://github.com/woocommerce/woocommerce-android/blob/trunk/.buildkite/pipeline.yml#L71-L74), as you see in this PR diff. I believe what @pachlava wanted to do in this PR is adding the same plugin to another test step, but with a different api-token-env-var value. See https://github.com/woocommerce/woocommerce-android/pull/9380#discussion_r1254645343

Ah, I missed that detail, thanks for pointing it out!

> I've used export BUILDKITE_ANALYTICS_TOKEN=$BUILDKITE_ANALYTICS_TOKEN_INSTRUMENTED_TESTS inside run-instrumented-tests.sh, but it always resulted in Missing BUILDKITE_ANALYTICS_TOKEN environment variable, because, as I see it, the var will be set in the shell script process, not in Buildkite command process.

I think your understanding is correct. The main difference with WCiOS is that WCiOS doesn't use the Test Collector as a Buildkite plugin, but as an SPM library. As such, the Test Analytics collection is done via the library, that can access the env var exposed by the shell script; while here for Android we want to use the `plugins: [*test_collector]`, where the buildkite plugin is the one using the token, so that's why it needs the env var to be declared at this level.

> I could not get this notation working (tbh, I don't understand what should I use for mytool), could you please explain it in more details, @AliSoftware? Thank you!

Sorry for the confusion, that was meant to be a dummy example for the general case of passing secret env variables to any `command` you might want to run as part of a pipeline `step`, without any specific tool in mind here. Like, say you had a script called `upload_to_app_center.sh` that needed an App Center token for it to work, then you should never use `command: .buildkite/command/upload_to_app_center.sh $APP_CENTER_TOKEN` from within your `pipeline.yml` — because that `$APP_CENTER_TOKEN` would be interpolated directly from the `pipeline.yml` and thus get exposed — but instead move that `$APP_CENTER_TOKEN` within your `.buildkite/command/upload_to_app_center.sh` script for it to be interpolated and evaluated within the shell script, and not within the `pipeline.yml`. That's what the "anti-pattern" documented in Buildkite is all about.

I guess that now that I take a closer look at your specific setup here with WCAndroid that uses a plugin, that dummy example was maybe not the best, since you don't pass the env var to your shell script or `command`, but instead want to pass it to the `test-collector` plugin. The warning still holds though about not using `$BUILDKITE_ANALYTICS_TOKEN` (with the `$`, that would make that env var be evaluated and expanded) from within your `pipeline.yml`.

---

> you don't pass the env var to your shell script or command, but instead want to pass it to the `test-collector` plugin. 

This is the case @AliSoftware. If I understand it right, it's passing per se, it's defining the value for `BUILDKITE_ANALYTICS_TOKEN` in the context of an individual step, which is then used as a default value for `api-token-env-var` option for the `test-collector` plugin.

> The warning still holds though about not using `$BUILDKITE_ANALYTICS_TOKEN` (with the `$`, that would make that env var be evaluated and expanded) from within your `pipeline.yml`.

Yes, this was my main concern here. I'm not using it with `$` anywhere, however, was unsure if I should be worried about using `$` for the variables from secrets.